src/modes/ModeEstimation.jl

```diff
@@ -0,0 +1,318 @@
+using ..Turing
+using ..Bijectors
+using LinearAlgebra
+
+import ..AbstractMCMC: AbstractSampler
+import ..DynamicPPL
+import ..DynamicPPL: Model, AbstractContext, VarInfo, AbstractContext, VarName,
+    _getindex, getsym, getfield, settrans!,  setorder!,
+    get_and_set_val!, istrans, tilde, dot_tilde
+import Optim
+import Optim: optimize
+import NamedArrays
+import ..ForwardDiff
+import StatsBase
+import Printf
+
+export MAP, MLE, optimize
+
+struct MLE end
+struct MAP end
+
+"""
+    OptimizationContext{C<:AbstractContext} <: AbstractContext
+
+The `OptimizationContext` transforms variables to their constrained space, but
+does not use the density with respect to the transformation. This context is
+intended to allow an optimizer to sample in R^n freely.
+"""
+struct OptimizationContext{C<:AbstractContext} <: AbstractContext
+	context::C
+end
+
+# assume
+function DynamicPPL.tilde(ctx::OptimizationContext{<:LikelihoodContext}, spl, dist, vn::VarName, inds, vi)
+    if haskey(vi, vn)
+        # Always overwrite the parameters with new ones for `SampleFromUniform`.
+        if spl isa SampleFromUniform || is_flagged(vi, vn, "del")
+            unset_flag!(vi, vn, "del")
+            r = init(dist, spl)
+            vi[vn] = vectorize(dist, r)
+            settrans!(vi, false, vn)
+            setorder!(vi, vn, get_num_produce(vi))
+        else
+            r = vi[vn]
+        end
+    else
+        r = init(dist, spl)
+        push!(vi, vn, r, dist, spl)
+        settrans!(vi, false, vn)
+    end
+	return r, 0
+end
+
+function DynamicPPL.tilde(ctx::OptimizationContext, spl, dist, vn::VarName, inds, vi)
+    if haskey(vi, vn)
+        # Always overwrite the parameters with new ones for `SampleFromUniform`.
+        if spl isa SampleFromUniform || is_flagged(vi, vn, "del")
+            unset_flag!(vi, vn, "del")
+            r = init(dist, spl)
+            vi[vn] = vectorize(dist, r)
+            settrans!(vi, false, vn)
+            setorder!(vi, vn, get_num_produce(vi))
+        else
+            r = vi[vn]
+        end
+    else
+        r = init(dist, spl)
+        push!(vi, vn, r, dist, spl)
+        settrans!(vi, false, vn)
+    end
+    return r, Distributions.logpdf(dist, r)
+end
+
+# observe
+function DynamicPPL.tilde(ctx::OptimizationContext{<:PriorContext}, sampler, right, left, vi)
+    return 0
+end
+
+function DynamicPPL.tilde(ctx::OptimizationContext, sampler, dist, value, vi)
+    return Distributions.logpdf(dist, value)
+end
+
+# dot assume
+function DynamicPPL.dot_tilde(ctx::OptimizationContext{<:LikelihoodContext}, sampler, right, left, vn::VarName, _, vi)
+    vns, dist = get_vns_and_dist(right, left, vn)
+    r = get_and_set_val!(vi, vns, dist, sampler)
+    return r, 0
+end
+
+function DynamicPPL.dot_tilde(ctx::OptimizationContext, sampler, right, left, vn::VarName, _, vi)
+    vns, dist = get_vns_and_dist(right, left, vn)
+    r = get_and_set_val!(vi, vns, dist, sampler)
+    lp = sum(logpdf(dist, r))
```

Comment:
I guess this should be either
```suggestion
    lp = sum(logpdf.(dist, r))
```
or
```suggestion
    lp = loglikelihood(dist, r)
```
IIRC `logpdf(dist, r)` where `r` are multiple samples is deprecated in Distributions. I'd prefer the latter since it avoid the creation of an intermediate array if not needed (e.g., without AD) but one could still implement custom adjoints based on broadcasting. It seems, however, that we currently always use the first alternative in DynamicPPL, so I guess the custom adjoints are not available right now.